cpp/include/cudf/strings/regex/regex_program.hpp

```diff
@@ -0,0 +1,138 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/strings/regex/flags.hpp>
+#include <cudf/types.hpp>
+
+#include <memory>
+#include <string>
+
+namespace cudf {
+namespace strings {
+
+/**
+ * @addtogroup strings_regex
+ * @{
+ */
+
+/**
+ * @brief Regex program class
+ *
+ * Create an instance from a regex pattern and use it to call the appropriate
+ * strings APIs. An instance can be reused.
+ *
+ * See the @ref md_regex "Regex Features" page for details on patterns and APIs the support regex.
+ */
+struct regex_program {
+  struct regex_program_impl;
+
+  /**
+   * @brief Create a program from a pattern
+   *
+   * @throw cudf::logic_error If pattern is found to be invalid or contain unsupported features
+   *
+   * @param pattern Regex pattern
+   * @param flags Regex flags for interpreting special characters in the pattern
+   * @param capture Controls how capture groups in the pattern are used
+   * @return Instance of this object
+   */
+  static std::unique_ptr<regex_program> create(std::string_view pattern,
+                                               regex_flags flags      = regex_flags::DEFAULT,
+                                               capture_groups capture = capture_groups::EXTRACT);
+
+  /**
+   * @brief Move constructor
+   *
+   * @param other Object to move from
+   */
+  regex_program(regex_program&& other);
+
+  /**
+   * @brief Move operator assignment
+   *
+   * @param other Object to move from
+   * @return this object
+   */
+  regex_program& operator=(regex_program&& other);
+
+  /**
+   * @brief Return the pattern used to create this instance
+   *
+   * @return regex pattern as a string
+   */
+  std::string pattern() const;
+
+  /**
+   * @brief Return the regex_flags used to create this instance
+   *
+   * @return regex flags setting
+   */
+  regex_flags flags() const;
+
+  /**
+   * @brief Return the capture_groups used to create this instance
+   *
+   * @return capture groups setting
+   */
+  capture_groups capture() const;
+
+  /**
+   * @brief Return the number of instructions in this instance
+   *
+   * @return Number of instructions
+   */
+  int32_t instructions_count() const;
+
+  /**
+   * @brief Return the number of capture groups in this instance
+   *
+   * @return Number of groups
+   */
+  int32_t groups_count() const;
+
+  /**
+   * @brief Return implementation object
+   *
+   * @return impl object instance
+   */
+  regex_program_impl* get_impl() const;
+
+  /**
+   * @brief Return the pattern used to create this instance
+   *
+   * @param num_strings Number of strings for computation
+   * @return Size of the working memory in bytes
+   */
+  std::size_t compute_working_memory_size(int32_t num_strings) const;
+
+  ~regex_program();
+
+ private:
+  regex_program();
+
+  std::string _pattern;
+  regex_flags _flags;
+  capture_groups _capture;
+
+  std::unique_ptr<regex_program_impl> _impl;
```

Comment:
Does this object need to be mutable? I think it can be constructed in the initializer list for `regex_program::regex_program` but I could be mistaken. To be clear I'm mostly thinking of `std::unique_ptr<regex_program_impl const>` (not a const pointer).

Reply:
Yes, I could not create the object on the initializer list.

Reply:
You could use an IIFE in the initializer list to make it work, but I don't know if it's a good thing to make the `_impl` immutable anyway.

Reply:
Adding a constructor(s) makes this work and I can now call `std::make_unique` in the ctor initializer list.